### ISSUE TITLE
Assembly version vs. file version of system.net.http

### ISSUE THREAD

**Issue by yohny** [microsoft/dotnet#1114]
We have a project targeting .Net framework 461, that is referencing System.Net.Http nuget (v4.3.4) which contains System.Net.Http.dll 4.1.1.3. That was the assembly actually being used (= ending up in build output folder).
Recently we added a reference to a netstandard2.0 library, which resulted in System.Net.Http.dll 4.2.0.0 being used instead (with proper assembly redirect being generated during build because of `<AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>` in .csproj).
So far, all is good.
Our project uses an installer (WIX toolset) and here comes the issue.
Installer updates/replaces the components/files based on their file version - aka. file is updated/replaced only if it has higher version than the one present. That seems to not be the case for System.Net.Http.dll, because "newer" assembly (4.2.0.0) has for some reason lower file version than "older" one (4.1.1.3).

Assembly version | File version
--------| ------------
4.1.1.3 | 4.6.26907.1
4.2.0.0 | 4.6.26011.1

Is this by design? Can not this cause issues with other tooling that uses file version?

**Comment by vivmishra** [comment]
@davidsh  @scalablecory

**Comment by davidsh** [comment]
@ericstj  Can you please advise here?  Why is WIX installer using Fileversion instead of Assembly version?

Short answer:  Do not use the System.Net.Http NuGet package.  And on .NET Framework, it is unnecessary since the required binary for System.Net.Http.dll is already part of the .NET Framework.

I understand in your case that you didn't intentionally reference System.Net.Http NuGet package. It came as a result of referencing netstandard2.0.  Usually the auto binding redirects will help.

**Comment by ericstj** [comment]
All installers use fileversion when doing upgrades.  It is by design that some assembly version and file version may not be in sync.  We use asssembly version to map to the API version, it may be different depending on the framework the assembly applies to, whereas fileversion is time-based.

**Comment by yohny** [comment]
Just to clarify our case: the nuget dependency to System.Net.Http was not added "by us", it was added as secondary dependency of some other nuget we use directly (long before we added reference to netstandard lib).
When we added netstandard2.0 library reference, suddenly different system.net.http.dll appeared in build output (we did not do any changes other that referencing the mentioned netstandard2.0 lib)

Not using system.net.http nuget is a bit tricky, as when you add nuget that depends on System.Net.Http, it is added as nuget, you have no control over that
Whats the proper workflow then? Whenever you add a nuget, you should manually check if some additionally added nugets dont have their framework counterpart and use that instead? That seems like something tooling (nuget/msbuild) should do...

What I dont get is why adding netstandard library reference has silent side-effect of changing used system.net.http from nuget assembly to framework assembly... Could not framework assembly be used from the beginnnig?

@ericstj  "fileversion is time-based" does that mean that 4.2.0.0 was created before 4.1.1.3?

**Comment by ericstj** [comment]
Yes, it looks like we shipped the System.Net.Http package from the `release/1.1` branch a couple times to fix issues in the .NETCore implementation that don't apply to .NETFramework, so we never built this binary which comes from the `release/2.0` branch.  To address this we need to rebuild these packages from a servicing branch and pick those up in the SDK.

Can you workaound this in your installer by scheduling an early RemoveExistingProducts (EG: afterInstallInitialize)?  Or just tell users who hit it to delete the file and repair your product?

**Comment by yohny** [comment]
I'll try to work around this in installer as you suggest.
We are planning to update to .net framework 4.8 so I wonder If that will not solve the issue - what system.net.http assembly will end up being used then?

**Comment by ericstj** [comment]
Updating to 4.8 is even better.  Netstandard2.0 support is inbox here so you don’t need to bring System.Net.Http in your app at all.

**Comment by yohny** [comment]
I found another assembly that has the same issue: System.Security.Cryptography.X509Certificates.dll

Assembly version | File version
-- | --
4.1.1.2 | 4.6.25815.3
4.1.2.0 | 4.6.25714.1

to the installer workaround - scheduling `RemoveExistingProducts` early as `afterInstallInitialize` does not help - the result is that files with lower version are completely removed from installation directory as during the uninstall phase they are removed and install phase refuses to add them (log entry from MSI: `Disallowing installation of component: {GUID} since the same component with higher versioned keyfile exists`)

so until we upgrade the targeted .net framework the only workaround for users is to uninstall the product prior to installation of new version (well there is [REINSTALLMODE](https://docs.microsoft.com/en-us/windows/win32/msi/reinstallmode), but that is strongly discouraged)
